drivers/adc/adc_nrfx_saadc.c

```diff
@@ -26,170 +24,242 @@ static struct driver_data m_data = {
+		return -EINVAL;
+	}
+
+	err = gain_set(ch_cfg, channel_cfg->gain);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	err = reference_set(ch_cfg, channel_cfg->reference);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	err = acq_time_set(ch_cfg, channel_cfg->acquisition_time);
+	if (err != 0) {
+		return -EINVAL;
+	}
+
+	ch_cfg->mode = (channel_cfg->differential ?
+		NRF_SAADC_MODE_DIFFERENTIAL : NRF_SAADC_MODE_SINGLE_ENDED);
+
+	nrfx_saadc_channel_config(&cfg);
 
 	return 0;
 }
 
 static void adc_context_start_sampling(struct adc_context *ctx)
 {
-	nrf_saadc_enable(NRF_SAADC);
+	nrfx_err_t err;
 
 	if (ctx->sequence.calibrate) {
-		nrf_saadc_task_trigger(NRF_SAADC,
-				       NRF_SAADC_TASK_CALIBRATEOFFSET);
-	} else {
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_START);
-		nrf_saadc_task_trigger(NRF_SAADC, NRF_SAADC_TASK_SAMPLE);
+		nrfx_saadc_offset_calibrate(NULL);
+	}
+
+	err = nrfx_saadc_mode_trigger();
+
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("Cannot start sampling: %d", err);
 	}
 }
 
 static void adc_context_update_buffer_pointer(struct adc_context *ctx,
 					      bool repeat)
 {
+	nrf_saadc_value_t *buffer;
+	uint32_t buffer_size;
+
 	ARG_UNUSED(ctx);
 
-	if (!repeat) {
-		nrf_saadc_buffer_pointer_set(
-			NRF_SAADC,
-			nrf_saadc_buffer_pointer_get(NRF_SAADC) +
-			nrf_saadc_amount_get(NRF_SAADC));
+	if (repeat) {
+		buffer = m_data.event->data.done.p_buffer;
+		buffer_size = m_data.event->data.done.size;
+	} else {
+		buffer = m_data.event->data.done.p_buffer +
+			 m_data.event->data.done.size;
+		buffer_size = m_data.event->data.done.size;
 	}
+
+	nrfx_saadc_buffer_set(buffer, buffer_size);
 }
 
-static int set_resolution(const struct adc_sequence *sequence)
+static int get_resolution(const struct adc_sequence *sequence,
+			  nrf_saadc_resolution_t *resolution)
 {
-	nrf_saadc_resolution_t nrf_resolution;
-
 	switch (sequence->resolution) {
 	case  8:
-		nrf_resolution = NRF_SAADC_RESOLUTION_8BIT;
+		*resolution = NRF_SAADC_RESOLUTION_8BIT;
 		break;
 	case 10:
-		nrf_resolution = NRF_SAADC_RESOLUTION_10BIT;
+		*resolution = NRF_SAADC_RESOLUTION_10BIT;
 		break;
 	case 12:
-		nrf_resolution = NRF_SAADC_RESOLUTION_12BIT;
+		*resolution = NRF_SAADC_RESOLUTION_12BIT;
 		break;
 	case 14:
-		nrf_resolution = NRF_SAADC_RESOLUTION_14BIT;
+		*resolution = NRF_SAADC_RESOLUTION_14BIT;
 		break;
 	default:
 		LOG_ERR("ADC resolution value %d is not valid",
-			    sequence->resolution);
+			sequence->resolution);
 		return -EINVAL;
 	}
 
-	nrf_saadc_resolution_set(NRF_SAADC, nrf_resolution);
 	return 0;
 }
 
-static int set_oversampling(const struct adc_sequence *sequence,
-			    uint8_t active_channels)
+static int get_oversampling(const struct adc_sequence *sequence,
+			    uint8_t active_channels,
+			    nrf_saadc_oversample_t *oversampling)
 {
-	nrf_saadc_oversample_t nrf_oversampling;
-
 	if ((active_channels > 1) && (sequence->oversampling > 0)) {
 		LOG_ERR(
 			"Oversampling is supported for single channel only");
```

Comment:
this statement is wrong, as oversampling works correctly for multi-channel configuration provided that burst is enabled as well, and this is exactly what `nrfx_saadc_sample_mode_set` does.

Therefore this whole check should be removed

Reply:
@jakub-uC bump